third_party/pyth/evm-watcher/src/listener.ts

```diff
@@ -32,31 +43,135 @@ export class Listener {
+      if (this.batchPrices.has(batchHash) === false) {
+        this.batchPrices.set(batchHash, [])
+      }
+      this.batchPrices.get(batchHash)!.push(priceFeedInfo);
     });
-    const pythEmitter = this.pythContract.events.PriceUpdate(undefined, async (_error: Error, event: EventData) => {
-      console.dir(event);
+
+    const batchPriceFeedUpdateEmitter = this.pythContract.events.BatchPriceFeedUpdate(undefined, async (_error: Error, event: EventData) => {
+      const batchValues: BatchPriceFeedUpdateEventValues = {
+        chainId: event.returnValues.chainId,
+        sequenceNumber: event.returnValues.sequenceNumber,
+        batchSize: event.returnValues.batchSize,
+        freshPricesInBatch: event.returnValues.freshPricesInBatch,
+      };
+
+      const batchHash = getBatchHash(batchValues.chainId, batchValues.sequenceNumber);
+      if (this.batchPrices.has(batchHash) === false) {
+        this.batchPrices.set(batchHash, [])
+      }
+
+      const batchSize = Number(batchValues.batchSize)
+      const waitForPricesCondition = () => this.batchPrices.has(batchHash) && this.batchPrices.get(batchHash)!.length == batchSize;
+      const result = await waitForCondition(waitForPricesCondition, 5000);
+      if (result === false) {
+        console.error(`The price list for ${batchHash} didn't complete. Reason: Timeout`);
+        return;
+      }
+
+      const batchInfo: BatchPriceFeedInfo = {
+        priceUpdates: this.batchPrices.get(batchHash)!,
+        ...batchValues,
+      }
+
+      const txHash = event.transactionHash;
+      if (this.txBatches.has(txHash) === false) {
+        this.txBatches.set(txHash, [])
+      }
+      this.txBatches.get(txHash)!.push(batchInfo);
     });
+
+    const UpdatePriceFeedsEmitter = this.pythContract.events.UpdatePriceFeeds(undefined, async (_error: Error, event: EventData) => {
+      const updateValues: UpdatePriceFeedsEventValues = {
+        sender: event.returnValues.sender,
+        batchCount: event.returnValues.batchCount,
+        fee: event.returnValues.fee,
+      };
+
+      const txHash = event.transactionHash;
+
+      if (this.txBatches.has(txHash) === false) {
+        this.txBatches.set(txHash, [])
+      }
+
+      const batchCount = Number(updateValues.batchCount);
+      const waitForBatchesCondition = () => this.txBatches.has(txHash) && this.txBatches.get(txHash)!.length == batchCount;
+      const waitForBatchesResult = await waitForCondition(waitForBatchesCondition, 5000);
+      if (waitForBatchesResult === false) {
+        console.error(`Couldn't fetch all batches in transaction ${txHash}. Reason: Timeout`);
+        return;
+      }
+
+      const updatePriceFeedsInfo: UpdatePriceFeedsInfo = {
+        batchUpdates: this.txBatches.get(txHash)!,
+        ...updateValues,
+      }
+
+      const waitForTimestampCondition = () => this.blockToTimestamp.has(event.blockNumber);
+      const waitForTimestampResult = await waitForCondition(waitForTimestampCondition, 5000);
```

Comment:
i don't love this chained waiting logic but I can't think of a better way to do it, so I guess this is fine.

Reply:
Me neither, I will add an issue for the future to also solve bad assumption that I made. 

The solution I'm thinking about is parsing the transaction logs manually (with events abi) instead. This way no chaining is needed, but it's less clean as we need to match log to event ourselves.